File: server/src/services/stats_service.ts
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

async function dashboard(): Promise<{
    users: { total: number; todayNew: number };
    quotes: { total: number; free: number; memberOnly: number };
    collections: { total: number; quotes: number };
    todayActivity: {
        newOrders: number;
        newCollections: number;
        newRecitePlans: number;
        reviews: number;
    };
}> {
    const now = new Date();
    const startOfToday = startOfDay(now);

    const [
        totalUsers,
        todayNewUsers,
        totalQuotes,
        freeQuotes,
        memberOnlyQuotes,
        totalCollections,
        totalCollectionQuotes,
        newOrders,
        newCollections,
        newRecitePlans,
        reviews,
    ] = await Promise.all([
        prisma.user.count(),
        prisma.user.count({ where: { createdAt: { gte: startOfToday } } }),
        prisma.quote.count({ where: { isActive: true, auditStatus: 1 } }),
        prisma.quote.count({ where: { isActive: true, auditStatus: 1, isFree: true } }),
        prisma.quote.count({ where: { isActive: true, auditStatus: 1, isFree: false } }),
        prisma.collection.count(),
        prisma.collectionQuote.count(),
        prisma.order.count({ where: { createdAt: { gte: startOfToday } } }),
        prisma.collection.count({ where: { createdAt: { gte: startOfToday } } }),
        prisma.recitePlan.count({ where: { createdAt: { gte: startOfToday } } }),
        prisma.reciteLog.count({ where: { reviewDate: { gte: startOfToday } } }),
    ]);

    return {
        users: { total: totalUsers, todayNew: todayNewUsers },
        quotes: {
            total: totalQuotes,
            free: freeQuotes,
            memberOnly: memberOnlyQuotes,
        },
        collections: {
            total: totalCollections,
            quotes: totalCollectionQuotes,
        },
        todayActivity: {
            newOrders,
            newCollections,
            newRecitePlans,
            reviews,
        },
    };
}

/**
 * 首页公开统计数据（无需登录）
 * 返回真实的数据库统计值，用于用户首页卡片展示
 */
async function homeStats(): Promise<{
    quoteTotal: number;
    categoryTotal: number;
    templateTotal: number;
    dailyRecommendCount: number;
}> {
    const today = startOfDay(new Date());
    const tomorrow = new Date(today.getTime() + 24 * 60 * 60 * 1000);

    const [quoteTotal, categoryTotal, templateTotal, dailyCount] = await Promise.all([
        prisma.quote.count({ where: { isActive: true, auditStatus: 1 } }),
        prisma.category.count({ where: { status: 1 } }),
        prisma.compositionTemplate.count({ where: { status: 1 } }),
        prisma.dailyRecommendQuote.count({
            where: { recommend: { recommendDate: { gte: today, lt: tomorrow } } },
        }),
    ]);

    return {
        quoteTotal,
        categoryTotal,
        templateTotal,
        dailyRecommendCount: dailyCount,
    };
}

function startOfDay(d: Date): Date {
    const x = new Date(d);
    x.setHours(0, 0, 0, 0);
    return x;
}

export { dashboard, homeStats };

File: server/app/services/stats_service.py
"""统计服务（对应 TS stats_service.ts）"""
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from ..models import (
    Category,
    Collection,
    CollectionQuote,
    CompositionTemplate,
    DailyRecommend,
    DailyRecommendQuote,
    Order,
    Quote,
    ReciteLog,
    RecitePlan,
    User,
)


def _startOfDay(d: datetime) -> datetime:
    return d.replace(hour=0, minute=0, second=0, microsecond=0)


def dashboard(db: Session) -> dict:
    startOfToday = _startOfDay(datetime.now())

    totalUsers = db.query(User).count()
    todayNewUsers = db.query(User).filter(User.createdAt >= startOfToday).count()
    totalQuotes = db.query(Quote).filter(Quote.isActive == True, Quote.auditStatus == 1).count()
    freeQuotes = (
        db.query(Quote).filter(Quote.isActive == True, Quote.auditStatus == 1, Quote.isFree == True).count()
    )
    memberOnlyQuotes = (
        db.query(Quote).filter(Quote.isActive == True, Quote.auditStatus == 1, Quote.isFree == False).count()
    )
    totalCollections = db.query(Collection).count()
    totalCollectionQuotes = db.query(CollectionQuote).count()
    newOrders = db.query(Order).filter(Order.createdAt >= startOfToday).count()
    newCollections = db.query(Collection).filter(Collection.createdAt >= startOfToday).count()
    newRecitePlans = db.query(RecitePlan).filter(RecitePlan.createdAt >= startOfToday).count()
    reviews = db.query(ReciteLog).filter(ReciteLog.reviewDate >= startOfToday).count()

    return {
        "users": {"total": totalUsers, "todayNew": todayNewUsers},
        "quotes": {
            "total": totalQuotes,
            "free": freeQuotes,
            "memberOnly": memberOnlyQuotes,
        },
        "collections": {
            "total": totalCollections,
            "quotes": totalCollectionQuotes,
        },
        "todayActivity": {
            "newOrders": newOrders,
            "newCollections": newCollections,
            "newRecitePlans": newRecitePlans,
            "reviews": reviews,
        },
    }


def homeStats(db: Session) -> dict:
    today = _startOfDay(datetime.now())
    tomorrow = today + timedelta(days=1)

    quoteTotal = db.query(Quote).filter(Quote.isActive == True, Quote.auditStatus == 1).count()
    categoryTotal = db.query(Category).filter(Category.status == 1).count()
    templateTotal = db.query(CompositionTemplate).filter(CompositionTemplate.status == 1).count()
    dailyCount = (
        db.query(DailyRecommendQuote)
        .join(DailyRecommend, DailyRecommendQuote.recommendId == DailyRecommend.id)
        .filter(
            DailyRecommend.recommendDate >= today,
            DailyRecommend.recommendDate < tomorrow,
        )
        .count()
    )

    return {
        "quoteTotal": quoteTotal,
        "categoryTotal": categoryTotal,
        "templateTotal": templateTotal,
        "dailyRecommendCount": dailyCount,
    }
